Restructure guide "singing txs offline" - Closes #1315 (#1330)
LiskHQ/lisk-docs: modules/ROOT/pages/guides/node-management/signing-transactions-offline.adoc

= Creating transactions and signing them offline
:toc: preamble
:idprefix:
:idseparator: -
:url_guides_decoding: guides/app-development/decoding-transactions-blocks.adoc
:url_transfer_asset: references/lisk-framework/token-module.adoc#transferasset
:url_register_multisig_asset: references/lisk-framework/keys-module.adoc#registerasset
:url_register_delegate_asset: references/lisk-framework/dpos-module.adoc#registertransactionasset
:url_vote_delegate_asset: references/lisk-framework/dpos-module.adoc#votetransactionasset
:url_unlock_asset: references/lisk-framework/dpos-module.adoc#unlocktransactionasset
:url_pom_asset: references/lisk-framework/dpos-module.adoc#pomtransactionasset
:url_framework_httpapi: references/lisk-framework/http-api-plugin.adoc
:url_service: lisk-service::index.adoc

> How to create and sign a transaction in a completely offline environment.

Reviewer: ```suggestion
How to create and sign a transaction in an entirely offline environment.
```

== Creating transactions

=== Creating transactions with lisk-client

==== Require the transaction and asset schema

> To create, validate and sign transactions offline with the lisk-client, it is necessary to access their schemas.

Reviewer: ```suggestion
To create, validate, and sign transactions offline with the lisk-client, it is necessary to access their schemas.
```

> Create a new file `schemas.js` and add all required schemas in this file, like shown in the snippet below.

Reviewer: ```suggestion
Create a new file `schemas.js` and add all required schemas in this file, as shown in the snippet below.
```

> In this example, we will send a Token Transfer transaction, therefore we need two different schemas:

Reviewer: ```suggestion
In this example, we will send a Token Transfer transaction, therefore two different schemas are required:
```

--nonce=2 --offline --json --pretty
? Please enter: amount:  1000000000
? Please enter: recipientAddress:  ab0041a7d3f7b2c290b5b834d46bdc7b7eb85815
? Please enter: data:  send tokens
? Please enter passphrase:  [hidden]
? Please re-enter passphrase:  [hidden]
{
  "transaction": "0802100018022080c2d72f2a20e03c09bdc8c023d94cf66a5d352e6258380210d97d545abbf75668ea3736e3123229088094ebdc031214ab0041a7d3f7b2c290b5b834d46bdc7b7eb858151a0b73656e6420746f6b656e733a40faa2626d7306506b1999f48aa2f4b1ffdee01e641fa76d37a9d1d6fd8c225a81065c856ea625c52d138a7e3ba86b62913dc8e5aef8b5e307641ab66e0277a60b"
}
{
  "transaction": {
    "moduleID": 2,
    "assetID": 0,
    "nonce": "2",
    "fee": "100000000",
    "senderPublicKey": "e03c09bdc8c023d94cf66a5d352e6258380210d97d545abbf75668ea3736e312",
    "signatures": [
      "faa2626d7306506b1999f48aa2f4b1ffdee01e641fa76d37a9d1d6fd8c225a81065c856ea625c52d138a7e3ba86b62913dc8e5aef8b5e307641ab66e0277a60b"
    ],
    "asset": {
      "amount": "1000000000",
      "recipientAddress": "ab0041a7d3f7b2c290b5b834d46bdc7b7eb85815",
      "data": "send tokens"
    }
  }
}
----
====

> == Broadcasting/ sending a signed transaction

Reviewer: ```suggestion
== Broadcasting/sending a signed transaction
```

A transaction can be posted to a node in the following ways:

* <<via-api-client>>
* <<via-rpc-api>>
* <<via-cli>>
* <<via-lisk-service-http-api>>
* <<via-lisk-core-http-api>>

[NOTE]
====

> If the existing data needs to encoded/ decoded before posting, check the guide xref:{url_guides_decoding}[]

Reviewer: ```suggestion
If the existing data needs to be encoded/decoded before posting, check the guide xref:{url_guides_decoding}[]
```

//Use public Service node
//const WS_RPC_ENDPOINT = "wss://service.lisk.com/rpc-v2";

// 2. Connect to Lisk Service via WebSockets
const socket = io(WS_RPC_ENDPOINT, {
  forceNew: true,
  transports: ['websocket']
});

// 3. Emit the remote procedure call
socket.emit('request', {
  jsonrpc: '2.0',
  method: 'post.transactions',
  payload: {"transaction":"08021000180d2080c2d72f2a200fe9a3f1a21b5530f27f87a414b549e79a940bf24fdf2b2f05e7f22aeeecc86a32270880c2d72f12144fd8cc4e27a3489b57ed986efe3d327d3de40d921a0a73656e6420746f6b656e3a4069242925e0e377906364fe6c2eed67f419dfc1a757f73e848ff2f1ff97477f90263487d20aedf538edffe2ce5b3e7601a8528e5cd63845272e9d79c294a6590a"}
},
  answer => {
    // console.log(answer);
    jsome(answer);
    process.exit(0);
});
----

=== Via CLI

[source,bash]
----
$ lisk-core transaction:send 0802100018022080c2d72f2a20e03c09bdc8c023d94cf66a5d352e6258380210d97d545abbf75668ea3736e3123229088094ebdc031214ab0041a7d3f7b2c290b5b834d46bdc7b7eb858151a0b73656e6420746f6b656e733a40faa2626d7306506b1999f48aa2f4b1ffdee01e641fa76d37a9d1d6fd8c225a81065c856ea625c52d138a7e3ba86b62913dc8e5aef8b5e307641ab66e0277a60b
----

=== Via Lisk Service HTTP API

> IMPORTANT: xref:{url_service}[Lisk Service] needs to be installed in oder to post a transaction via the Lisk Service HTTP API.

Reviewer: ```suggestion
IMPORTANT: xref:{url_service}[Lisk Service] needs to be installed in order to post a transaction via the Lisk Service HTTP API.
```

> cURL is one of the tools that can be used to send HTTP API requests to Lisk Service:

Reviewer: ```suggestion
cURL is one of the tools that can be used to send HTTP API requests to Lisk Service as shown below:
```

[source,bash]
----
curl -X POST -H "Content-Type: application/json" \
-d '{ "transaction": "0802100018022080c2d72f2a20e03c09bdc8c023d94cf66a5d352e6258380210d97d545abbf75668ea3736e3123229088094ebdc031214ab0041a7d3f7b2c290b5b834d46bdc7b7eb858151a0b73656e6420746f6b656e733a40faa2626d7306506b1999f48aa2f4b1ffdee01e641fa76d37a9d1d6fd8c225a81065c856ea625c52d138a7e3ba86b62913dc8e5aef8b5e307641ab66e0277a60b"}' \
"http://localhost:9901/api/v2/transactions"
----

> The following response will be displayed, if the transaction was posted successfully.

Reviewer: ```suggestion
The following response shown below will be displayed if the transaction was posted successfully:
```

----
{"message":"Transaction payload was successfully passed to the network node","transactionId":"8a503843942e7d47ba0bef83fe735d26381f32a6ca6c96fb1cde902315f6220c"}
----

=== Via Lisk Core HTTP API

> IMPORTANT: The xref:{url_framework_httpapi}[] needs to be enabled on the node in oder to post a transaction via the node HTTP API.

Reviewer: ```suggestion
IMPORTANT: The xref:{url_framework_httpapi}[] needs to be enabled on the node in order to post a transaction via the node HTTP API.
```

> cURL is one of the tools that can be used to send HTTP API requests to Lisk Core:

Reviewer: ```suggestion
cURL is one of the tools that can be used to send HTTP API requests to Lisk Core as shown below:
```

[source,bash]
----
curl -X POST -H "Content-Type: application/json" \
-d '{ "transaction": "0802100018022080c2d72f2a20e03c09bdc8c023d94cf66a5d352e6258380210d97d545abbf75668ea3736e3123229088094ebdc031214ab0041a7d3f7b2c290b5b834d46bdc7b7eb858151a0b73656e6420746f6b656e733a40faa2626d7306506b1999f48aa2f4b1ffdee01e641fa76d37a9d1d6fd8c225a81065c856ea625c52d138a7e3ba86b62913dc8e5aef8b5e307641ab66e0277a60b"}' \
"http://localhost:4000/api/transactions"
----

> The following response will be displayed, if the transaction was posted successfully.

Reviewer: ```suggestion
The following response shown below will be displayed if the transaction was posted successfully:
```

Author: ```suggestion
The following response will be displayed if the transaction was posted successfully:
```
